src/VisualStudio/Core/Impl/CodeModel/CodeModelTaskExtensions.cs

```diff
@@ -2,24 +2,24 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using System.Threading;
 using System.Threading.Tasks;
-using Roslyn.Utilities;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
 
 namespace Microsoft.VisualStudio.LanguageServices.Implementation.CodeModel
 {
     internal static class CodeModelTaskExtensions
     {
         /// <summary>
-        /// Does a Roslyn.Utilities.TaskExtensions.WaitAndGetResult{T} for CodeModel.
+        /// Easy extension for running a task synchronously for CodeModel code.
         /// </summary>
         /// <remarks>
-        /// This function is the exact same as Roslyn.Utilities.TaskExtensions.WaitAndGetResult{T}. CodeModel, since it
-        /// must implement a highly blocking API, has no choice but to use WaitAndGetResult in a bunch of places. We
-        /// hope to audit _CanCallOnBackground periodically, and so rather than having to understand that each of those
-        /// uses are CodeModel and thus get a special pass.
+        /// CodeModel only exists in Visual Studio, and is required to be a synchronous API. Since it may
+        /// now call into other processes, which then could call back to the UI thread in VS, we need a non
+        /// UI blocking way to handle this behavior. This forces the use of JTF as the way to do asynchronous
+        /// work synchronously in the CodeModel layer. It is not advised that this is general purpose to all
+        /// of Roslyn editor layer.
         /// </remarks>
-        public static T WaitAndGetResult_CodeModel<T>(this Task<T> task, CancellationToken cancellationToken)
-            => task.WaitAndGetResult(cancellationToken);
+        public static T WaitAndGetResult_CodeModel<T>(this Task<T> task, IThreadingContext threadingContext)
+            => threadingContext.JoinableTaskFactory.Run(() => task);
```

Comment:
> `.Run(() => task)`

❗ This is only valid for JTF if `task` is a `JoinableTask<T>`.

Reply:
wait, really? How does code normally do things like `_threadingContext.JoinableTaskFactory.Run(() =>document.GetCompilationAsync())` for example? I'm not sure I see the difference. 

Should I check for `JoinableTask<T>` directly and use JTF, otherwise use normal `WaitAndGetResult`?

Reply:
> How does code normally do things like _threadingContext.JoinableTaskFactory.Run(() =>document.GetCompilationAsync()) for example?

The `Run` method configures values in the execution context which change the behavior of the callback function. If the task is created in advance, it is not known to have been created in an execution context with the correct values.

Reply:
> The Run method configures values in the execution context which change the behavior of the callback function. If the task is created in advance, it is not known to have been created in an execution context with the correct values.

Gotcha, that makes sense. It would mean that if I did add an async lambda and await the task it would be fine right? The thread would yield correctly while executing the awaited task. `Run(async () => await task.ConfigureAwait(false);)`

> There isn't a need to change WaitAndGetResult_CodeModel. For code that wants to use JTF, that code should use JTF directly instead of calling WaitAndGetResult_CodeModel

The idea here was that it's not easy to know whether to use JTF or not. It's not obvious that `Renamer.RenameAsync` for example would. I'm not even sure I know of all the places that would require it. I can just do the specific change for now if that seems better.

Reply:
I went ahead and made the change to `async () => await task`. Let me know what you think and if we should just do surgical changes instead of this. I'm in favor of a "I don't have to think about it" approach, but I don't know how much churn we expect in this code.

Reply:
In order to use JTF at this location, the argument would need to be changed from `Task<T>` to `Func<Task<T>>`. Then all uses like this:

```
MethodAsync().WaitAndGetResult_CodeModel()
```

would be changed to this:

```
WaitAndGetResult_CodeModel(() => MethodAsync())
```

However, this change would deviate from the general rule that JTF.Run should not be wrapped in other helpers. If we wanted to use JTF for all cases where `WaitAndGetResult_CodeModel` is used today, there is exactly one allowed option:

1. Delete `WaitAndGetResult_CodeModel`
2. Replace all code that used `WaitAndGetResult_CodeModel` with a call to `threadingContext.JoinableTaskFactory.Run`

Reply:
Got it. I really wish JTF had a `WaitAndGetResult` like extension...

Reply:
It has `Join`, which can be used for this purpose provided the task was originally started with `RunAsync`. The `Run` method is effectively these two operations in sequence.